bittide/src/Bittide/Transceiver/WordAlign.hs

```diff
@@ -0,0 +1,178 @@
+  , joinMsbs
+  , dealignBytes
+  ) where
+
+import Clash.Prelude
+
+import Bittide.SharedTypes (Bytes)
+import Clash.Explicit.Prelude (noReset)
+import Data.Bifunctor (Bifunctor (bimap))
+import Data.Maybe (fromMaybe)
+import Data.Tuple.Extra (curry3)
+
+data State n m = State { prev :: BitVector (n * m), offset :: Index n }
+  deriving (Show, Generic, ShowX, NFDataX)
+
+defState :: (KnownNat n, KnownNat m) => State n m
+defState = State{prev=0, offset=0}
+
+alignSymbol :: forall n . (KnownNat n, 1 <= n) => BitVector n
+alignSymbol = bit (natToNum @(n - 1))
+
+-- | Split the MSBs of a 'BitVector's \"bytes\" into a 'BitVector' of MSBs and a
+-- 'BitVector' of the remaining bits.
+splitMsbs ::
+  forall nBytes byteWidth .
+  ( KnownNat nBytes
+  , KnownNat byteWidth
+  , 1 <= byteWidth
+  ) =>
+  BitVector (nBytes * byteWidth) ->
+  ( BitVector nBytes
+  , BitVector (nBytes * (byteWidth - 1))
+  )
+splitMsbs =
+    bimap pack pack
+  . unzip
+  . unpack @(Vec nBytes (Bit, BitVector (byteWidth - 1)))
+
+-- | Opposite of 'splitMsbs'.
+joinMsbs ::
+  forall nBytes byteWidth .
+  ( KnownNat nBytes
+  , KnownNat byteWidth
+  , 1 <= byteWidth
+  ) =>
+  BitVector nBytes ->
+  BitVector (nBytes * (byteWidth - 1)) ->
+  BitVector (nBytes * byteWidth)
+joinMsbs msbs bvs = pack $
+  zip
+    (unpack @(Vec nBytes Bool) msbs)
+    (unpack @(Vec nBytes (BitVector(byteWidth - 1))) bvs)
+
+-- | Specialized version of 'align' that works on 'BitVector' and assumes that
+-- the alignment bits are stored in the MSBs of the bytes.
+alignBytesFromMsbs ::
+  forall n dom .
+  ( HiddenClock dom
+  , KnownNat n
+  , 1 <= n
+  ) =>
+  -- | Freeze alignment
+  Signal dom Bool ->
+  -- | Data with alignment bits in the byte's MSBs
+  Signal dom (Bytes n) ->
+  -- | 'Nothing' until an alignment symbol has been observed. 'Just' if an alignment
+  -- symbol has been observed.
+  Signal dom (Bytes n)
+alignBytesFromMsbs freeze dat =
+  let msbs = fst . splitMsbs @n <$> dat
+   in align freeze (unpack <$> msbs) dat
+
+-- | Align data according to an alignment signal. The alignment signal is a
+-- one-hot encoded signal that indicates which bit is the alignment bit. Note
+-- that the behavior of this function is undefined when:
+--
+--   * Switching from one alignment bit to another alignment bit
+--   * No alignment bit or multiple are asserted
+--
+-- Alignment can be "frozen" by asserting the freeze signal. This is useful when
+-- switching over from a handshake protocol to a data protocol.
+align ::
+  forall n m dom .
+  ( HiddenClock dom
+  , KnownNat n
+  , KnownNat m
+  , 1 <= n
+  ) =>
+  -- | Freeze alignment
+  Signal dom Bool ->
+  -- | One hot encoded alignment signal. Behavior is undefined when multiple
+  -- bits or none are asserted.
+  Signal dom (Vec n Bool) ->
+  -- | Data
+  Signal dom (BitVector (n * m)) ->
+  -- | 'Nothing' until an alignment symbol has been observed. 'Just' if an
+  -- alignment symbol has been observed.
+  Signal dom (BitVector (n * m))
```

Comment:
Same mismatch in the documentation here.

Reply:
Yes. See other comment.